docs/tools/test_spec_gen.py

```diff
@@ -0,0 +1,126 @@
+from typing import Dict
+
+from doc_generator import ClassVisitor, FuncVisitor, extract_metadata
+
+base_path = Path(__file__).parent
+file_path = (base_path / "../run_test/test_spec.rst").resolve()  # path of test_spec.rst
+test_path = (
+    base_path / "../../microsoft/testsuites"
+).resolve()  # where tests are present
+
+
+def update_file(filename: Path = file_path) -> None:
+    """
+    Updates (rewrites) test table.
+
+    Args:
+        filename (Path, optional): path of the table. Defaults to file_path.
+    """
+    with open(file_path, "r+") as test_spec:
+        write_title(test_spec)
+
+        for root, _, files in os.walk(test_path):
+            for file in files:
+                if file.endswith(".py"):
+                    print("Processing " + file)
+                    filename = Path(root) / file
+                    with open(filename, "r") as f:
+                        contents = f.read()
+                        tree = ast.parse(contents)
+                    c = ClassVisitor()
+                    f = FuncVisitor()
+                    c.visit(tree)
+                    f.visit(tree)
+
+                    for suite in extract_metadata(c.classes):
+                        # case["case_name"] = case["name"]
+                        # del case["name"]
+                        write_suite(test_spec, suite)
+                        for case in extract_metadata(f.functions):
+                            # suite["suite_name"] = suite["name"]
+                            # del suite["name"]
+                            write_case(test_spec, case)
+
+
+def write_title(file: TextIOWrapper) -> None:
+    title = "Test Specification"
+    file.write(title + "\n")
+    file.write("=" * len(title) + "\n")
+    file.write("\n")
+
+    file.write("This file lists all test cases' specifications.\n")
+    file.write("\n")
+
+
+def write_suite(file: TextIOWrapper, metadata: Dict[str, str]) -> None:
+    """
+    A helper function to write info of a test suite.
+
+    Args:
+        file (TextIOWrapper): test spec file
+        metadata (Dict[str, str]): test suite metadata
+    """
+    # file.write("\n")
+    file.write(".. class:: ")
+    file.write(metadata["name"] + "\n")  # Test Suite Name
+    file.write("    :noindex:" + "\n")
+    file.write("\n")
+
+    # TODO
+    file.write("    :description: TODO")  # Description
+    file.write("\n")
+
+    file.write("    :platform: ")
+    file.write("Azure, Ready, Local" + "\n")  # Platform
+
+    file.write("    :area: ")
+    file.write(metadata["area"] + "\n")  # Area
+
+    file.write("    :category: ")
+    file.write(metadata["category"] + "\n")  # Category
+    file.write("\n")
+
+
+def write_case(file: TextIOWrapper, metadata: Dict[str, str]) -> None:
+    """
+    A helper function to write info of a test case.
+
+    Args:
+        file (TextIOWrapper): test spec file
+        metadata (Dict[str, str]): test case metadata
+    """
+    file.write("    .. method:: ")
```

Comment:
Try to use a template file, it's easier to update. In the Python file, it generates the key parts only. Or generate a data array, and parse it in the html template.

Reply:
Not sure how to use a html template. Needs more knowledge. 
IMO I can do refactor afterwards to make sure the rst labels are easy to modify.

Reply:
A template is to save all lines like header and styles in a file. The code composites some parts, and then replace keywords in the template file.

Reply:
If it's high cost, we can skip this change.

Reply:
I'm trying to figure it out how to make a template. I will let you know if it would cost too much.